### DISCUSSION TITLE
[Java] How should we handle "loggers"?

### DISCUSSION THREAD

**Discussion by chrisdutz** [apache/plc4x#973] [Polls]
When going through our code, I've noticed that we have different ways of defining our loggers. I think it would be cool, if we could decide on one.

In general there are two ways to declare the loggers:
1. as a local variable
2. as a static variable

However usually static variables are all uppercase, but I'm also seeing some static loggers named as local variables.

**Comment by chrisdutz** [comment]
And here ... even the Slf4j folks no longer recommend any general way of doing things: https://www.slf4j.org/faq.html#declared_static
I guess in the past, there was a performance reason to to things, but this no longer seems to apply today.

**Comment by sruehl** [comment]
With the up-rise of structured loggers, context became more relevant. So with that in mind it is pretty helpful to have loggers enriched and pass it down to the downstream objects which enrich them by themselfs with their context variables.
Recently in golang I moved to this pattern out of 3 reasons:

1. Have more control over the context
2. Easier way to configure them
3. Being able to assign those log statements to the relevant tests

Point 3 becomes painfully important once you move into async territory

So I would vote for having them instance bound.

**Reply by chrisdutz** [reply]
Good point ... so there's a "almost not noticeable overhead" compared with a lot of new options, right?

**Comment by chrisdutz** [comment]
Would be awesome, if we could see who voted how :-/

**Comment by foxpluto** [comment]
I would suggest for Java to use the notation `@Log4j2` present in Lombok to decrease the boilerplate an configure the logging notation externally as described here: [@Log (and friends)](https://projectlombok.org/features/log).

I voted for "log"

**Reply by chrisdutz** [reply]
Well admittedly I would be hesitant to use lombok as it requires our users to setup additional plugins in their development environments.

**Reply by sruehl** [reply]
I think in most IDEs it is enabled by default nowadays

**Reply by foxpluto** [reply]
@chrisdutz  It depends, if Lombok will  be used for more feature than the simple logs one, enabling the plugin in the IDE will be worth. Lombok is a great tool for  killing the Boiler Plate.

**Reply by glcj** [reply]
Hello,

I think we don't use loombok in the project (pom.xml).

I think we should not add more libraries than necessary if it can be solved within the JDK tools.

My grain of sand

**Reply by sruehl** [reply]
not having to write Boilerplate is also preventing bugs like writing getters and returning the wrong property...

**Reply by chrisdutz** [reply]
Well you don't have to write boilerplate anyway as the IDE generates that for me .... So I do prefer that over having some tool generate stuff magically.

**Reply by sruehl** [reply]
a3ea768d2f7b35211d553505c9509b2c28a000d2 ([here](https://github.com/apache/plc4x/commit/a3ea768d2f7b35211d553505c9509b2c28a000d2#diff-29f99c8cd536dee245291d0c43c1db428fc4c628e4111670fb10d567c3ce18e1L188))
